Makefile

```diff
@@ -2,37 +2,148 @@
+	@printf -- "\n"
+	@printf -- "---------------------------------------------------------------------------------------\n"
+	@printf -- "Nix user? You can use ${FORMATTING_BEGIN_YELLOW}\`direnv allow .\`${FORMATTING_END} or ${FORMATTING_BEGIN_YELLOW}\`nix-shell --pure\`${FORMATTING_END}\n"
+	@printf -- "Want to use ${FORMATTING_BEGIN_YELLOW}\`docker\`${FORMATTING_END} or ${FORMATTING_BEGIN_YELLOW}\`podman\`${FORMATTING_END}? See ${FORMATTING_BEGIN_YELLOW}\`ENVIRONMENT=true\`${FORMATTING_END} commands. (Default ${FORMATTING_BEGIN_YELLOW}\`CONTAINER_TOOL=docker\`${FORMATTING_END})\n"
+	@printf -- "\n"
+	@awk 'BEGIN {FS = ":.*##"; printf "Usage: make ${FORMATTING_BEGIN_BLUE}<target>${FORMATTING_END}\n"} /^[a-zA-Z0-9_-]+:.*?##/ { printf "  ${FORMATTING_BEGIN_BLUE}%-46s${FORMATTING_END} %s\n", $$1, $$2 } /^##@/ { printf "\n\033[1m%s\033[0m\n", substr($$0, 5) } ' $(MAKEFILE_LIST)
+
+##@ Environment
+# We use a volume here as non-Linux hosts are extremely slow to share disks, and Linux hosts tend to get permissions clobbered.
+define ENVIRONMENT_EXEC
+	@echo "Entering environment..."
+	@mkdir -p target
+	$(CONTAINER_TOOL) run \
+			--name vector-environment \
+			--rm \
+			$(if $(findstring true,$(ENVIRONMENT_TTY)),--tty,) \
+			--init \
+			--interactive \
+			--env INSIDE_ENVIRONMENT=true \
+			--network host \
+			--mount type=bind,source=${PWD},target=/vector \
+			--mount type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock \
+			--mount type=volume,source=vector-target,target=/vector/target \
+			--mount type=volume,source=vector-cargo-cache,target=/root/.cargo \
+			$(ENVIRONMENT_UPSTREAM)
+endef
+
+define ENVIRONMENT_COPY_ARTIFACTS
+	@echo "Copying artifacts off volumes... (Docker errors below are totally okay)"
+	@mkdir -p ./target/release
+	@mkdir -p ./target/debug
+	@mkdir -p ./target/criterion
+	@$(CONTAINER_TOOL) rm -f vector-build-outputs || true
+	@$(CONTAINER_TOOL) run \
+		-d \
+		-v vector-target:/target \
+		--name vector-build-outputs \
+		busybox true
+	@$(CONTAINER_TOOL) cp vector-build-outputs:/target/release/vector ./target/release/ || true
+	@$(CONTAINER_TOOL) cp vector-build-outputs:/target/debug/vector ./target/debug/ || true
+	@$(CONTAINER_TOOL) cp vector-build-outputs:/target/criterion ./target/criterion || true
+	@$(CONTAINER_TOOL) rm -f vector-build-outputs
+endef
+
+
+ifeq ($(ENVIRONMENT_AUTOBUILD), true)
+define ENVIRONMENT_PREPARE
+	@echo "Building the environment. (ENVIRONMENT_AUTOBUILD=true) This may take a few minutes..."
+	$(CONTAINER_TOOL) build \
+		$(if $(findstring true,$(VERBOSE)),,--quiet) \
+		--tag $(ENVIRONMENT_UPSTREAM) \
+		--file scripts/environment/Dockerfile \
+		.
+endef
+else
+define ENVIRONMENT_PREPARE
+	$(CONTAINER_TOOL) pull $(ENVIRONMENT_UPSTREAM)
+endef
+endif
+
 
-##@ Default
+environment: ## Enter a full Vector dev shell in $CONTAINER_TOOL, binding this folder to the container.
+	${ENVIRONMENT_PREPARE}
+	@export ENVIRONMENT_TTY=true
```

Comment:
Note that this may cause surprises in a parallel make environment. Does it need to be set for all recipe lines that run after it?

Reply:
You're definitely right, if someone depended on this target as a dependency it would be an issue. I'll localize it better!